src/library/scala/concurrent/ExecutionContext.scala

```diff
@@ -139,7 +139,25 @@ object ExecutionContext {
    *
    * @return the global `ExecutionContext`
    */
-  def global: ExecutionContextExecutor = Implicits.global.asInstanceOf[ExecutionContextExecutor]
+  final def global: ExecutionContextExecutor = Implicits.global.asInstanceOf[ExecutionContextExecutor]
+
+  /**
+   * The `callingThread` `ExecutionContext` executes `Runnable`s on the calling `Thread`.
+   * Nested submissions will be trampolined to prevent uncontrolled stack space growth.
+   * Any `NonFatal` or `InterruptedException`s will be reported to the `defaultReporter`.
+   *
+   * In order to maximize chances of maintaining application responsiveness, it is strongly
+   * advised to only execute logic which will quickly return control to the caller.
+   *
+   * DO NOT call any blocking code in the `Runnable`s submitted to this `ExecutionContext`
+   * as it will prevent progress by other enqueued `Runnable`s and the calling `Thread`.
+   */
+  final object callingThread extends ExecutionContextExecutor with BatchingExecutor {
+    override final def submitForExecution(runnable: Runnable): Unit = runnable.run()
+    override final def execute(runnable: Runnable): Unit = submitSyncBatched(runnable)
+    override final def reportFailure(t: Throwable): Unit =
```

Comment:
Should we add explicit memory fences so that this execution context establishes the happens-before relationships that the normal one would?

Reply:
I don't think it's necessary.

There's no need for a memory barrier since it's guaranteed that the callback will run on the calling thread, so there is an intrinsic happens before relationship there. So, as long as the callback in the calling threads state is appropriately synchronised (which this executor can't do anything about, memory barrier or not, if it's not), then everything should be fine.

So if we then think about this in the context of the Scala Future API, the `DefaultPromise` implementation uses `AtomicReference.compareAndSet` to add and remove callbacks to be executed. There is a memory barrier here (eg, on x86 its implemented using `CMPXCHG` which gives a memory barrier), so any state held by callbacks that then eventually get submitted to this executor will have their state synchronised so as to guarantee happens before before being submitted to the executor. Other implementations of `Future` likewise must implement a memory barrier on all their callback capturing methods, if they don't it doesn't matter what the executor does, they won't be able to guarantee happens before between the caller of the method (eg the caller of `map`), and the execution of the callback to that method.

Reply:
@retronym @jroper Yes, since the execution does not leave the calling thread then there is no need for any fences.